File: tests/benchmarks/cpp/03_fibonacci_iterative.cpp
// ベンチマーク3: フィボナッチ数列（反復版）
// 1,000,000番目のフィボナッチ数を計算

#include <chrono>
#include <iostream>

using namespace std;
using namespace std::chrono;

int fibonacci_iter(int n) {
    if (n <= 1)
        return n;

    int prev = 0;
    int curr = 1;
    const int MOD = 1000000007;

    for (int i = 2; i <= n; i++) {
        int next = (prev + curr) % MOD;
        prev = curr;
        curr = next;
    }

    return curr;
}

int main() {
    int n = 1000000;

    auto start = high_resolution_clock::now();
    int result = fibonacci_iter(n);
    auto end = high_resolution_clock::now();

    auto duration = duration_cast<microseconds>(end - start);

    cout << "Fibonacci(" << n << ") mod 1000000007 = " << result << endl;
    cout << "Time: " << duration.count() / 1000000.0 << " seconds" << endl;

    return 0;
}

File: tests/benchmarks/rust/03_fibonacci_iterative.rs
// ベンチマーク3: フィボナッチ数列（反復版）
// 1,000,000番目のフィボナッチ数を計算

use std::time::Instant;

fn fibonacci_iter(n: i32) -> i32 {
    if n <= 1 {
        return n;
    }

    let mut prev = 0;
    let mut curr = 1;
    const MOD: i32 = 1000000007;

    for _ in 2..=n {
        let next = (prev + curr) % MOD;
        prev = curr;
        curr = next;
    }

    curr
}

fn main() {
    let n = 1000000;

    let start = Instant::now();
    let result = fibonacci_iter(n);
    let duration = start.elapsed();

    println!("Fibonacci({}) mod 1000000007 = {}", n, result);
    println!("Time: {:.6} seconds", duration.as_secs_f64());
}